### PR TITLE
UI beamline design

### PR DESCRIPTION
**Pull request by EnricoAhlers**
This implements the first version of a beamline design functionality in RAYX-UI.
Every element in the beamline can be selected, and its parameters can be changed.

Not yet supported:
- Some surface parameters (e.g. Quadric with its 16 parameters)
- Groups
- Saving RMLs
- Adding elements and changing the order

### CODE REVIEW

File: Intern/rayx-ui/src/UserInterface/BeamlineDesignHandler.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    void showBeamlineDesignWindow(UIBeamlineInfo& uiBeamlineInfo);
+
+  private:
+    void showParameters(DesignMap& parameters, bool& changed, uint32_t type);
+    void createInputField(const std::string& key, DesignMap& element, bool& changed, uint32_t type, int nestingLevel);
+
+    bool caseInsensitiveCompare(const std::string& a, const std::string& b);
+
+    // Define groups
+    std::map<std::string, std::vector<std::string>> groups = {
```
Comment (Atraxus) [inline Intern/rayx-ui/src/UserInterface/BeamlineDesignHandler.h:17]: The map should be const

File: Intern/rayx-ui/src/UserInterface/BeamlineDesignHandler.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+         {"additionalOrder", "DesignAlphaAngle", "DesignBetaAngle", "DesignEnergy", "DesignMeridionalEntranceArmLength",
+          "DesignMeridionalExitArmLength", "DesignOrderDiffraction", "DesignSagittalEntranceArmLength", "DesignSagittalExitArmLength",
+          "FresnelZOffset", "imageType"}},
+        {"Surface", {"geometricalShape", "behaviourType",    "curvatureType",     "CutoutLength",   "CutoutWidth",
+                     "Material",         "totalLength",      "lineDensity",       "deviationAngle", "radius",
+                     "OrderDiffraction", "lognRadius",       "shortRadius",       "armLength",      "parameter_a11",
+                     "parameter_P",      "parameter_P_type", "entranceArmLength", "exitArmLength",  "designGrazingIncAngle",
+                     "figureRotation",   "longHalfAxisA",    "shortHalfAxisB",    "longRadius"}}};
+
+    std::vector<std::string> customOrder = {"type",
```
Comment (Atraxus) [inline Intern/rayx-ui/src/UserInterface/BeamlineDesignHandler.h:45]: This should be a const array of const strings

File: Intern/rayx-ui/src/UserInterface/BeamlineOutliner.cpp [hunk 3]
```diff
@@ -42,14 +78,14 @@ void BeamlineOutliner::buildTreeFromXMLNode(rapidxml::xml_node<>* node, TreeNode
 
         if (strcmp(xmlChild->name(), "object") == 0) {
             if (type == "Point Source" || type == "Matrix Source" || type == "Dipole" || type == "Dipole Source" || type == "Pixel Source" ||
```
Comment (Atraxus) [inline Intern/rayx-ui/src/UserInterface/BeamlineOutliner.cpp:80]: A switch case would be faster here

**Review by Atraxus — CHANGES_REQUESTED** [review]

File: Intern/rayx-ui/src/UserInterface/UIHandler.cpp [hunk 4]
```diff
@@ -354,9 +358,23 @@ void UIHandler::showSceneEditorWindow(UIParameters& uiParams) {
         ImGui::EndDisabled();
     }
 
+    ImGui::Separator();
+
+    m_BeamlineDesignHandler.showBeamlineDesignWindow(uiParams.beamlineInfo);
+
+    ImGui::End();
+
+    ImGui::End();
```
[the remark this answers is not in the document]
Comment (EnricoAhlers) [reply Intern/rayx-ui/src/UserInterface/UIHandler.cpp:367]: It crashes with one 😄 But I analyzed it and saw that the corresponding Begin() is in setupUI so I moved one of the End() there.

File: Intern/rayx-ui/src/UserInterface/BeamlineDesignHandler.h (hunk 2 above)
Comment (EnricoAhlers) [reply Intern/rayx-ui/src/UserInterface/BeamlineDesignHandler.h:45]: I made the vector const, but it can't hold const strings

**Pull request merged**